### PR TITLE
[ci] Add tests for CI

### PR DESCRIPTION
**Pull request by jigold**
The last test for making sure PRs get merged is still a work in progress.

### CODE REVIEW

File: batch/batch/batch.py [hunk 1]
```diff
@@ -314,10 +314,11 @@ def _keep_alive_container(self):  # pylint: disable=R0201
...
     async def _delete_pod(self):
-        err = await app['k8s'].delete_pod(name=self._pod_name)
-        if err is not None:
-            traceback.print_tb(err.__traceback__)
-            self.log_info(f'ignoring pod deletion failure due to {err}')
+        pass
+        # err = await app['k8s'].delete_pod(name=self._pod_name)
+        # if err is not None:
+        #     traceback.print_tb(err.__traceback__)
+        #     self.log_info(f'ignoring pod deletion failure due to {err}')
```
Comment (cseed) [inline batch/batch/batch.py]: Did you mean to leave this in?

File: batch/batch/k8s.py [hunk 2]
```diff
@@ -44,7 +44,8 @@ def __init__(self, blocking_pool, timeout, namespace, k8s_api):
 
     async def delete_pod(self, name):
         assert name is not None
-        return await self._delete_pod(name=name)
+        return None
+        # return await self._delete_pod(name=name)
```
Comment (cseed) [inline batch/batch/k8s.py]: Same.

File: build.yaml [hunk 3]
```diff
@@ -33,7 +33,10 @@ steps:
        valueFrom: default_ns.name
    public: false
    secrets:
+    - database-server-config
+    - gcr-push-service-account-key
```
Comment (cseed) [inline build.yaml]: So these are OK for now, since we only test authorized changes and protect the logs, but eventually we're going to want to fully isolate the tests from production.  You don't need to change anything, just something to keep in mind.

File: build.yaml [hunk 4]
```diff
@@ -33,7 +33,10 @@ steps:
...
     - test-gsa-key
+    - ci-gsa-key
```
Comment (cseed) [inline build.yaml]: I'm not sure why this is necessary.  auth/create-accounts creates a (test) ci-gsa-key.

File: build.yaml [hunk 5]
```diff
@@ -817,6 +838,53 @@ steps:
...
+     echo creating $REPO_NAME...
+     curl -XPOST \
+       -i \
+       https://api.github.com/orgs/hail-ci-test/repos \
+       -H "Authorization: token ${TOKEN}" \
+       -d "{ \"name\" : \"$REPO_NAME\" }"
+
+     # checkout hail code
+     cd /
+     mkdir -p repo/hail
```
Comment (cseed) [inline build.yaml]: Just `mkdir -p /repo/hail` since the next line is a `cd`.

File: ci/ci/build.py [hunk 6]
```diff
@@ -585,6 +593,9 @@ def build(self, batch, code, scope):  # pylint: disable=unused-argument
 date
 '''
 
+        if is_test_deployment:
+            script = 'true'
```
Comment (cseed) [inline ci/ci/build.py]: Isn't it easier to just skip the job an return at the top?

File: build.yaml [hunk 7]
```diff
@@ -52,6 +55,17 @@ steps:
...
+   name: deploy_ci_sa
+   namespace:
+     valueFrom: batch_pods_ns.name
+   config: ci/service-account.yaml
+   dependsOn:
+    - default_ns
+    - batch_pods_ns
+   scopes:
+    - test
+    - dev
```
Comment (cseed) [inline build.yaml]: This is the actual ci-agent configuration, right?  In which case, this can be unconditional so ci-agent is managed by build.yaml now, right?

File: batch/batch/batch.py (hunk 1 above)
Comment (jigold) [reply batch/batch/batch.py]: It's the only way I can get the logs for failing pods right now. I'll delete later. I have to keep this as a PR because it works in my namespace since batch-pods == default in dev deploy.

File: ci/ci/build.py (hunk 6 above)
Comment (jigold) [reply ci/ci/build.py]: I didn't know if we had to return a job if there's dependencies on that step.

File: build.yaml (hunk 7 above)
Comment (jigold) [reply build.yaml]: I didn't want to overwrite ci-agent in the batch-pods namespace until I knew what permissions it had. I can't figure out how to get the yaml configuration using kubectl and I also looked at the yaml configuration on the broad server.

File: ci/ci/build.py (hunk 6 above)
Comment (cseed) [reply ci/ci/build.py]: We don't return a job, we assign to the job attribute of steps.  By default it is None.  When computing dependencies, we add dependencies to all non-None jobs of parent steps.  So, in a word, nope!

File: build.yaml (hunk 7 above)
Comment (cseed) [reply build.yaml]: Ah, yeah, you can't because ci-agent has cluster-wide admin bindings because it needs to be able to create and destroy namespaces.

There are two kinds of roles and role binding: cluster wide and namespace local.  ci-agent has a cluster-wide role:

```
$ kubectl get -o yaml clusterrolebinding ci-agent-cluster-admin
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: ci-agent-cluster-admin
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: cluster-admin
subjects:
- kind: ServiceAccount
  name: ci-agent
  namespace: batch-pods
```

Comment (cseed) [reply build.yaml]: My suggestion was overly ambitious, disregard.

File: batch/batch/batch.py (hunk 1 above)
Comment (cseed) [reply batch/batch/batch.py]: Oh, yeah, sure, no worries.

File: build.yaml (hunk 4 above)
Comment (jigold) [reply build.yaml]: I can't figure out why either. For some reason the ci-gsa-key is trying to be mounted in the first BuildImage step. I get this error: `MountVolume.SetUp failed for volume "gsa-key" : secret "ci-gsa-key" not found`. I'll keep investigating because it doesn't make sense why this is happening.

Comment (jigold) [reply build.yaml]: It's in the batch1 code. We mount these volumes:

```
        volumes = [
            kube.client.V1Volume(
                secret=kube.client.V1SecretVolumeSource(
                    secret_name=self.userdata['gsa_key_secret_name']),
                name='gsa-key'),
            kube.client.V1Volume(
                secret=kube.client.V1SecretVolumeSource(
                    secret_name='batch-gsa-key'),
                name='batch-gsa-key')]

```

Comment (jigold) [reply build.yaml]: So the "user" gsa key needs to be in the namespace for this to work!

Comment (jigold) [reply build.yaml]: Oh, I see your concern. Let me think about this some more.

Comment (jigold) [reply build.yaml]: Ok. I fixed it. We needed to also copy the ci-gsa-key in create-accounts to the batch-pods namespace.

File: ci/test/test_ci.py [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+    yield gh_client
+    await session.close()
+
+
+async def test_deploy():
+    wait_for_hello()
+
+
+# FIXME: This test requires either putting user1 as an authorized user
+#        or having a fake developer who can authorize the sha
```
Comment (cseed) [inline ci/test/test_ci.py]: We should make the users authorized.

File: build.yaml [hunk 9]
```diff
@@ -52,6 +55,28 @@ steps:
...
+   namespace:
+     valueFrom: batch_pods_ns.name
+   config: ci/ci-agent-sa.yaml
+   dependsOn:
+    - batch_pods_ns
+   scopes:
+    - test
+    - dev
+ - kind: deploy
+   name: deploy_ci_agent_role
```
Comment (cseed) [inline build.yaml]: I'd merge these two deploys steps.  Makes things faster, and I find too much fragmentation in the yaml hard to follow.

File: build.yaml [hunk 10]
```diff
@@ -817,6 +849,53 @@ steps:
     - default_ns
     - base_image    
     - test_batch2
+ - kind: runImage
+   name: create_ci_test_repo
+   image:
+     valueFrom: base_image.image
+   script: |
+     set -e
```
Comment (cseed) [inline build.yaml]: Not having -x makes these scripts hard to debug, but there are insecure tokens in here.  What we've done elsewhere is turn on -x everywhere, and turn it off in the critical sections with an echo saying what's happening.  Something like:

```
set -ex

REPO_NAME="ci-test-{{token}}"

set +x
TOKEN=$(cat /secret/ci-secrets/user1)
echo creating $REPO_NAME...
curl -XPOST \
       -i \
       -sSL \
       https://api.github.com/orgs/hail-ci-test/repos \
       -H "Authorization: token ${TOKEN}" \
       -d "{ \"name\" : \"$REPO_NAME\" }"
set -x
...

Maybe not worth changing since this is working, but FYI.

File: ci/test/test_ci.py [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+    session = aiohttp.ClientSession(
+        raise_for_status=True,
+        timeout=aiohttp.ClientTimeout(total=60))
+
+    gh_client = gh_aiohttp.GitHubAPI(session, 'test-ci-2', oauth_token=user2_token)
+    yield gh_client
+    await session.close()
+
+
+async def test_deploy():
```
Comment (cseed) [inline ci/test/test_ci.py]: Does this need to be async?

File: ci/test/test_ci.py [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+import sys
+import pytest
+from gidgethub import aiohttp as gh_aiohttp
+import aiohttp
+import subprocess as sp
+import asyncio
+
+from hailtop.config import get_deploy_config
+
+pytestmark = pytest.mark.asyncio
```
Comment (cseed) [inline ci/test/test_ci.py]: I didn't know about this, this is great!

File: ci/test/test_ci.py (hunk 8 above)
Comment (cseed) [reply ci/test/test_ci.py]: And add other, unauthorized users for testing the authorize sha endpoint (tho that seems like a pretty low-priority test at this point).

File: ci/Dockerfile.test [hunk 13]
```diff
@@ -0,0 +1,14 @@
+FROM {{ base_image.image }}
+
+COPY hail/python/setup-hailtop.py /hailtop/setup.py
+COPY hail/python/hailtop /hailtop/hailtop/
+RUN python3 -m pip install --no-cache-dir /hailtop \
+  && rm -rf /hailtop
```
Comment (cseed) [inline ci/Dockerfile.test]: Since you need hailtop, you might just build off service_base_image which has hailtop, gear and web_common installed.
